### ISSUE TITLE
Support motion sensor K4659

### ISSUE THREAD

**Issue by superinj** [anotherjulien/MyHOME#18]
I have some of this sensor, there is a way to use it or must be coded?

I can tell you what frames they send:

- every time the motion is detected it send `*1*34*03##`, **WHAT 34** and **WHERE 03** A=0 PL=3
- every time it detect luminance change it send `*#1*03*6*2##`, **WHERE 03** A=0 PL=3, **DIMENSION 6**, **DIMENSION VALUE 2**

Do you think that you can add this feature?

Thank you

**Comment by anotherjulien** [comment]
Interesting prospect, yes.
I can look into it!

Do you know if there is any documentation on this? It's not in the OpenWebNet WHO=1 doc, but I did not expect it to be... 🙁 

WHAT 34 is motion detected, do you know what is sent for the "end" of the motion detection?
Do you have any idea of the rough scale of the luminance detection of DIMENSION 6? Or is it just binary for light/dark? (in that case, 2 is light, what is dark?)

**Comment by anotherjulien** [comment]
With values between 0 and 65534, it can conceivably be understood as Lux. Even if the values are off, it's also probably the easiest way to add it to HA.
I find it strange that the movement sensor would trigger only when movement is detected, the ones that act as dry contact (WHO=25) do also send a specific frame when the movement "stops"
Have you ever tried to capture the traffic on the bus for a while? Did you ever see a WHAT 33 or 35 for WHERE 03?

By the way, since this is on WHO 1, I suppose that if you have your sensor on WHERE 03, you cannot have a light on the same APL, right?

**Comment by superinj** [comment]
Yes dimension values may be interpreted as lux, though they are sometimes too low in my opinion to be real lux. 
I confirm you that STOP frame doesn't exist, I monitored it for a while but no other frames appear; in the configuration you can set how often the sensor should re-trigger, default is 15 min but you can set seconds if you want.

Yes I have my sensor in WHERE 03, but the sensor can be configured in different ways.
You can use it standalone with an its own APL ( and it sends the frames I wrote in the upper posts ("scenario" configuration); or you can bind it to a light with the same APL to turn light on or off ("light" configuration), in this case it follow standard frames of WHO 1.

I use the first one so I can do automations in HA.

**Comment by anotherjulien** [comment]
I was looking at the [datasheet](https://catalogue.bticino.com/app/webroot/low_res/487383_501678_ST_00000220_EN.pdf) for this sensor and I've seen it's suppose to detect light in the 20-1275 lux range.
(Or -maybe- you can set a detection threshold between 0 and 1275 lux to switch a light on, but it can detect a wider range than that?)
Do you know what is the lowest (non zero) value you've seen from it?
I suppose that a phone light directly hitting it straight on will max it out, so a reading of about 500 is likely to mean 1275 lx.
What do you think?

A problem I'm facing right now as well is that there is no integrated mechanism in HA to have an "auto-off" binary sensor.
So the only thing that comes to mind to achieve this would be to set automation to turn it off after a give time.
Do you know of any other way to do it?

**Comment by anotherjulien** [comment]
Actually, I've ordered a couple of 048822 sensors (which should work the same) along with a 088230 programming remote, so I'll test this myself as well when it arrives :-)
But already I can see that another documentation in French mentions that you need to calibrate the light sensors, so that might be why the values are off.

**Comment by superinj** [comment]
> Do you know what is the lowest (non zero) value you've seen from it?
> I suppose that a phone light directly hitting it straight on will max it out, so a reading of about 500 is likely to mean 1275 lx.
> What do you think?

I have one of this sensor in my garage, at night if I turn light on (very good illumination) it send dimension value 9, if I turn light off it send 65535.

> A problem I'm facing right now as well is that there is no integrated mechanism in HA to have an "auto-off" binary sensor.
> So the only thing that comes to mind to achieve this would be to set automation to turn it off after a give time.
> Do you know of any other way to do it?

Maybe you could treat it as an event? Send an event when motion is detected, than in your config anyone can do automation or a trigger based template sensor.

**Comment by superinj** [comment]
> Actually, I've ordered a couple of 048822 sensors (which should work the same) along with a 088230 programming remote, so I'll test this myself as well when it arrives :-)
> But already I can see that another documentation in French mentions that you need to calibrate the light sensors, so that might be why the values are off.

I 'm very happy you took it serously :smile:

**Comment by anotherjulien** [comment]
Would you be able to test the current master on GitHub?
Configuration is not yet documented, it should be as follows:

```
binary_sensor:
  - platform: myhome
    devices:
      office_motion:
        who: '1'
        where: '0312'
        name: Bureau
        class: motion
        manufacturer: Legrand
        model: 48822

sensor:
  - platform: myhome
    devices:
      office_illuminance:
        where: '0312'
        name: Bureau
        class: illuminance
        manufacturer: Legrand
        model: 48822
```

From my own tests, motion detection works fine and I use the sensor's configured timeout value to trigger the "end of motion detection"; so the binary sensor works as a true motion detection device.
Illuminance is indeed in lux (as per Legrand documentation, and by comparison with another lux meter)

Let me know how it works for you!

**Comment by superinj** [comment]
I tried it out, very good work! You surprised me with values from device config.
Motion sensors are working, timeout values that I get are as configured on device plus 15s, did you need to add this delta?
Illuminance sensors aren't showing in homeassistant, I configured it as you suggested, but I can't find it among entities. Something different that I could try?

**Comment by anotherjulien** [comment]
Normally, both the motion and illuminance sensors should be 2 entities for the same device:
![Screenshot 2021-09-13 at 20 44 57](https://user-images.githubusercontent.com/3992976/133139734-984188bf-7601-4b40-a9eb-2c3010318cbf.png)

Is there any errors in the logs?
Can you provide debug logs to see what happens?

**Comment by superinj** [comment]
I got it working, homeassistant wants a single file starting with this:

> sensor:
>  - platform: myhome
>    devices:

So I joined it with other one and it worked.
As I said in previous posts, one of this sensor (K4659) send a strange value at dark when light is off:

<img width="373" alt="Schermata 2021-09-14 alle 00 03 06" src="https://user-images.githubusercontent.com/1248711/133162792-082bbec2-293c-4e6c-8218-ffeb3faf3f54.png">

Can you exclude this single value? Have you tried to calibrate the sensor with his remote?

**Comment by anotherjulien** [comment]
I've never had to try the calibration because the values I get out of the box are well within the rage of other light meters in the same room.
![Screenshot 2021-09-14 at 01 26 50](https://user-images.githubusercontent.com/3992976/133169995-f0659d4d-1100-4bd4-8041-8ed8ba64059a.png)
I might need to when I put them in their final place since it will be embedded in the ceiling.

I've never faced the problem that seem very common for you to receive the 65535 lx value!
Do you have multiple sensors of the same type? Do they all produce the same behaviour?

If the issue is widespread I'd be more than happy to include a 'fix' even if it's a bit of a bodge, but this would be the wrong place to compensate for defective hardware in my opinion.

**Comment by superinj** [comment]
> Do you have multiple sensors of the same type? Do they all produce the same behaviour?
Unfortunately I have only one of this kind, the other ones are BMSE3001 (like yours I suppose).
THis sensor has thi trange value from the beginning, maybe I can order a new one and try it.

**Comment by anotherjulien** [comment]
If you want to test bypassing 65535, you can use the latest commit and uncomment lines 553 and 554 in `sensor.py`
If I can get feedback for other devices/users that this issue is common (for this specific value), I might consider pushing this to everyone.

**Comment by superinj** [comment]
I tested it with this mod and it works:
```
if message.illuminance == 65535:
    self._attr_native_value = 0
else:
    self._attr_native_value = message.illuminance
```
Non zero values seems right, I should measure it with a lux meter.
I think you can add it to the master because get 65535 as real value it's almost impossible and no one would be affected.

If you want you can close the issue, thank you and good work!

**Comment by anotherjulien** [comment]
What was the behaviour with the potential fix I included?
My understanding is that your proposed fix is as much of a problem, but maybe just less visible in the graphs. 0 is still not the real value.

I disagree that 65535 is almost impossible. The Legrand 48834 sensor is IP55 rated, and as such *can* be used outdoors (even if Legrand pretends to the contrary).
I don't live in a very sunny area and very often get more than 55000 lx on my outdoors Zigbee sensors. So I don't think 65000 lx is unthinkable in southern Europe.
Besides, at this point we have no indication that any other sensor than yours has the same problem.
Conversely, there could be sensors with the same problem but a different value.

**Comment by superinj** [comment]
With your fix the value was not updated, if the sensor send "10" with lights on and then "65535" when the lights turn off, with "return true" the sensor show "10 lx" until a new valid value is sent, and no trace that the lights were off.

I understand your doubts to include this fix, I think I should maintain a fork of your project to keep this fix just for me.

**Comment by anotherjulien** [comment]
I've published version 0.8.3 with support added, we'll see if the issue you have is reproductible with more users.
